lib/framedivider.rb

```diff
@@ -390,46 +422,284 @@ def opening(model, id)
-  vec << OpenStudio::Point3d.new(pts[:C][:p].x, pts[:C][:p].y, pts[:C][:p].z)
-  vec << OpenStudio::Point3d.new(pts[:D][:p].x, pts[:D][:p].y, pts[:D][:p].z) if four
-
-  return u, area, points unless fit?(model, id, vec)
-
-  tr1 = []
-  tr1 << pts[:A][:p]
-  tr1 << pts[:B][:p]
-  tr1 << pts[:C][:p]
-  area1 = areaHeron(tr1)
-
-  area2 = 0
-  if four
-    tr2 = []
-    tr2 << pts[:A][:p]
-    tr2 << pts[:C][:p]
-    tr2 << pts[:D][:p]
-    area2 = areaHeron(tr2)
-  end
-  area = area1 + area2
-
-  return u, area, vec
+  ptz
+end
+
+##
+# Fetch a surface's subsurface opening areas & vertices.
+#
+# @param [OpenStudio::Model::Model] model An OpenStudio model
+# @param [OpenStudio::Model::Surface] surface An OpenStudio surface
+#
+# @return [Hash] Returns surface with key attributes, including openings.
+def openings(model, surface)
+  surf = {}
+
+  unless model && model.is_a?(OpenStudio::Model::Model)
+    msg = "Can't find OpenStudio model for 'openings' - skipping"
+    TBD.log(TBD::DEBUG, msg)
+    return nil
+  end
+  unless surface && surface.is_a?(OpenStudio::Model::Surface)
+    msg = "Can't find OpenStudio surface for 'openings' - skipping"
+    TBD.log(TBD::DEBUG, msg)
+    return nil
+  end
+
+  identifier = surface.nameString
+
+  if surface.space.empty?
+    msg = "Can't find space holding '#{identifier}' - skipping"
+    TBD.log(TBD::ERROR, msg)
+    return nil
+  end
+  surf[:space] = surface.space.get
+
+  a = surf[:space].spaceType.empty?
+  surf[:stype] = surf[:space].spaceType.get unless a
+  a = surf[:space].buildingStory.empty?
+  surf[:story] = surf[:space].buildingStory.get unless a
+
+  # Site-specific (or absolute, or true) surface normal.
+  t, r = transforms(model, surf[:space])
+  unless t && r
+    msg = "Can't process '#{identifier}' space transformation"
+    TBD.log(TBD::FATAL, msg)
+    return nil
+  end
+
+  n = trueNormal(surface, r)
+  unless n
+    msg = "Can't process '#{identifier}' true normal"
+    TBD.log(TBD::FATAL, msg)
+    return nil
+  end
+  surf[:n] = n
+
+  surf[:gross] = surface.grossArea
+  fd = false
+  subs = {}
+
+  # Option A:
+  # subsurfaces = model.getSubSurfaces.sort_by{ |s| s.nameString }
+  # subsurfaces.each do |s|
+
+  # Option B:
+  # surface.subSurfaces.each do |s|
+
+  # Option C:
+  # model.getModelObjects.each do |s|
+    # next if s.to_SubSurface.empty?
+    # s = s.to_SubSurface.get
+
+  # Option D:
+  subsurfaces = surface.subSurfaces.sort_by { |s| s.nameString }
+  subsurfaces.each do |s|
+    id = s.nameString
+
+    unless s.vertices.size == 3 || s.vertices.size == 4
+      msg = "Subsurface '#{id}' vertex count (must be 3 or 4) - skipping"
+      TBD.log(TBD::ERROR, msg)
+      next
+    end
+
+    typ = s.subSurfaceType.downcase
+    type = :skylight
+    type = :window if typ.include?("window")
+    type = :door if typ.include?("door")
+    glazed = true if type == :door && typ.include?("glass")
```

Comment:
Are windows not glazed?

Reply:
Of course - always. However doors may or may not de glazed (and this does matter for some energy codes - including the Canadian ones). Here, we're just tagging a TBD `:door` as glazed (or not) based on the subsurface type "Door" vs "GlassDoor".